---
title: "DESEQ_analysis"
author: "Sascha Maschmann"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      fig.align = "center")

packages <- c("plotly", "tidyverse", "ggrepel", "circlize")

installed <- rownames(installed.packages())
to_install <- setdiff(packages, installed)

if (length(to_install) > 0) {
  message("Installing missing packages: ", paste(to_install, collapse = ", "))
  install.packages(to_install, repos = "https://cloud.r-project.org")
}

library(plotly)
library(tidyverse)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(edgeR)

```

## Analysis of mito reads

After finding suitable scaling genes, we use their total read numbers per sample normalize the mito mapped reads.  

```{r load data}
countFiles <- list.files("../../out/counts",
                         pattern = "*mito_featureCounts.txt$",
                         full.names = TRUE)

absoluteNormalizators <- base::readRDS("intermediates/absolute_norm_values.RDS") %>% select(-con_b)

coldata <- read.csv("../../samples.csv")

mitoCounts <- read.delim(countFiles[1], header = T, skip = 1) %>% 
  select(Geneid, last_col())
for (file in countFiles[2:length(countFiles)]){
  counts <- read.delim(file, header = T, skip = 1) %>% 
    select(Geneid, last_col())
  mitoCounts <- left_join(mitoCounts, counts, by = "Geneid")
}


# Filter out all genes with less than 10 reads in all samples
cleanCounts <- mitoCounts %>% 
  filter(if_all(-Geneid, ~ .x >= 25),
         Geneid != "RNA19ii")

colnames(cleanCounts) <- base::gsub("out.filtered.",
                                         "",
                                         colnames(cleanCounts))
colnames(cleanCounts) <- base::gsub("_mito_filtered.bam",
                                         "",
                                         colnames(cleanCounts))
colnames(cleanCounts) <- base::gsub("\\.",
                                         "_",
                                         colnames(cleanCounts))
```

```{r plot raw reads for example genes, fig.asp=9/19}
# Set example genes throughout the document
example_genes = c("SSUB", "RNA10", "RNA33")

example_plot <- cleanCounts %>% 
  filter(Geneid %in% example_genes) %>% 
  pivot_longer(cols = 2:16,names_to = "con") %>% 
  mutate(con = substr(con, 1,3)) %>% 
  group_by(Geneid, con)

ggplot(example_plot,
       aes(x = con,
           y = value,
           color = Geneid)) +
  geom_point(show.legend = FALSE) +
  theme_bw() +
  scale_y_log10() +
  labs(x = "Condition",
       y = expression(log[10]("Absolute Reads")),
       title = "Raw Reads per Condition") +
  facet_wrap(~Geneid)

```

```{r Normalize}
# mitoCounts: matrix of mito features (rows) x samples (columns), raw counts
mitoCounts <- cleanCounts %>%
  mutate(Geneid = paste0(Geneid, "XX")) %>%
  column_to_rownames("Geneid") 

# group: factor of conditions for each sample (length = ncol(mitoCounts))
group <- factor(c("parental", "parental", "parental",
                  "pr_0", "pr_0", "pr_0",
                  "pr_1", "pr_1", "pr_1",
                  "pr_2", "pr_2", "pr_2",
                  "pr_3", "pr_3", "pr_3"), levels = unique(coldata$con))

# anchor A_j:   numeric vector of length ncol(mitoCounts)
#               IMPORTANT: A_j must exclude mitochondrial reads; >0 in every sample
Aj <- absoluteNormalizators$value


# 1) Prepare DGEList with offset-only normalization (NO TMM)
Aj <- Aj * (1e6 / median(Aj))       # optional rescale for stability
y  <- DGEList(mitoCounts)

y$samples$lib.size <- Aj
y$samples$group <- group

# 2) Design with all groups (no intercept), pr_0 will be the baseline in contrasts
group <- factor(group, levels = c("parental","pr_0","pr_1","pr_2","pr_3"))
design <- model.matrix(~ 0 + group)
colnames(design) <- levels(group)

# 3) Dispersion + fit (QL pipeline)
y  <- estimateDisp(y, design)
fit <- glmQLFit(y, design)

# 4) Define contrasts: (first) - (pr_0)
CT <- makeContrasts(
  par_pr0 = parental - pr_0,
  pr0_pr1 = pr_1    - pr_0,   # logFC > 0 => pr_1 > pr_0
  pr0_pr2 = pr_2    - pr_0,
  pr0_pr3 = pr_3    - pr_0,
  levels = design
)

# 5) Run tests and collect results
run_one <- function(fit, contr_name) {
  tt <- glmQLFTest(fit, contrast = CT[, contr_name])
  out <- topTags(tt, n = Inf)$table
  out$contrast <- contr_name
  out
}

res_list <- lapply(colnames(CT), function(nm) run_one(fit, nm))
res <- do.call(rbind, res_list)

# res has logFC (relative to the anchor), PValue, FDR for each feature/bin and contrast
split(res, res$contrast) |>
  lapply(function(df) df[order(df$FDR), ]) -> by_contrast

res_clean <- res %>% 
  rownames_to_column("Geneid") %>% 
  mutate(Geneid = sub("XX.*", "", Geneid))
```

```{r plot normalized reads, fig.asp=9/19}

normReads <- data.frame(norm_counts <- cpm(y, normalized.lib.sizes = TRUE)) %>%
  rownames_to_column("Geneid") %>%
  mutate(Geneid = sub("XX", "", Geneid)) %>%
  pivot_longer(cols = 2:16,
               names_to = "con",
               values_to = "value") %>%
  mutate(con = substr(con, 1,3))

ggplot(filter(normReads, Geneid %in% example_genes),
       aes(x = con,
           y = value,
           color = Geneid)) +
  geom_point(show.legend = FALSE) +
  scale_y_log10() +
  theme_bw() +
  labs(x = "Condition",
       y = expression(log[10]("Normalized Reads")),
       title = "Normalized Reads per Condition") +
  facet_wrap(~Geneid)

```

The normalization smoothes the differences between replicates, but preserves the overall trend.

```{r Heatmap all Genes, fig.height=7, fig.width=3}

lfc_matrix <- res_clean %>% 
  select(Geneid, logFC, contrast) %>% 
  pivot_wider(id_cols = Geneid,
              names_from = contrast,
              values_from = logFC) %>% 
  column_to_rownames("Geneid") %>% 
  as.matrix()

col = colorRamp2(c(ceiling(max(lfc_matrix, na.rm = TRUE)), 0, floor(min(lfc_matrix, na.rm = TRUE))),
                 c("red", "white", "blue"))
heatmap_all <- Heatmap(lfc_matrix,
        column_title = "LFCs mito rRNA fragments",
        cluster_columns = FALSE,
        col = col,
        na_col = "grey20",
        name = "LFC",
        )
plot(heatmap_all)
```

Based on unfiltered log2fold changes, the general trend continues: over all, mito rRNA transcription presumably is dampened by the suppression of HPR4 transcription.

```{r filtered Heatmap, fig.height=7, fig.width=3}
p_cutoff <- 0.05
fdr_cutoff <- 0.05

lfc_matrix_minimal <- res_clean %>% 
  mutate(logFC = ifelse(PValue <= p_cutoff & FDR <= fdr_cutoff, logFC, NA)) %>% 
  select(Geneid, logFC, contrast) %>% 
  pivot_wider(id_cols = Geneid,
              names_from = contrast,
              values_from = logFC) %>% 
  filter(if_any(-Geneid, ~!is.na(.x))) %>% 
  select_if(~sum(!is.na(.)) > 0) %>% 
  column_to_rownames("Geneid") %>% 
  arrange(pr0_pr2, pr0_pr3) %>%
  as.matrix()

col = colorRamp2(c(2, 0, floor(min(lfc_matrix_minimal, na.rm = TRUE))),
                 c("red", "white", "blue"))
heatmap_sig <- Heatmap(lfc_matrix_minimal,
        column_title = "Significantly changed \n mito rRNA fragments",
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        col = col,
        na_col = "grey20",
        name = "LFC"
        )
plot(heatmap_sig)
```

When filtering based on adjusted p value (<= 0.05) only decreases in rRNA fragment expression remain.

```{r Volcano static}
volcano <- res_clean %>% 
  select(-logCPM, -`F`) %>% 
  mutate(y_var = -log10(FDR),
         color = ifelse(PValue <= p_cutoff & FDR <= fdr_cutoff, "purple", "grey"),
         label = ifelse(PValue <= p_cutoff & FDR <= fdr_cutoff, Geneid, "")
  )

# Plot volcano plots
volcano_0_2 <- ggplot(data = filter(volcano, contrast == "pr0_pr2"),
       aes(x = logFC,
           y = y_var)) +
  # Plot LFC against FDR
  geom_point(aes(color = color),
             show.legend = FALSE) +
  scale_color_manual(values = c("grey", "purple")) +
  # Add labels
  geom_text_repel(aes(label = label),
                  segment.color = 'purple',
                  min.segment.length = .1,
                  max.overlaps = 15,
                  size = 3,
                  force = 8,
                  force_pull = 0.2) +
  # Add cutoff indicators
  geom_hline(yintercept = -log10(fdr_cutoff),
             linetype = "dashed",
             color = "grey40") +
  geom_vline(xintercept = -1,
             linetype = "dashed",
             color = "grey40") +
  # Adjust titles
  labs(x = expression(log[2]("fold change")),
       y = expression(-log[10](padj)),
       title = "Day 0 vs Day 2") +
  # Set theme
  theme_bw() 

plot(volcano_0_2)
```

For day 3 comparisons, the plot was made interactive, since the points are too dense for labels. Hover the Points to see the gene.

```{r Volcano interactive}
# Due to the dense Distribution of samples, this plot is created interactively with plotly
volc1 <- ggplot(data = filter(volcano, contrast == "pr0_pr3"),
       aes(x = logFC,
           y = y_var,
           ID = Geneid)) +
  # Plot LFC against FDR
  geom_point(aes(color = color),
             show.legend = FALSE) +
  scale_color_manual(values = c("grey", "purple")) +
  # Add cutoff indicators
  geom_hline(yintercept = -log10(fdr_cutoff),
             linetype = "dashed",
             color = "grey40") +
  geom_vline(xintercept = -1,
             linetype = "dashed",
             color = "grey40") +
  # Adjust titles
  labs(x = expression(log[2]("fold change")),
       y = expression(-log[10](padj)),
       title = "Day 0 vs Day 3") +
  # Set theme
  theme_bw() 

suppressWarnings(ggplotly(volc1, tooltip = c("ID")))


```

```{r save plots, results='hide'}
png(filename = "plots/heatmap_all_LFC.png",
       width = 10,
       height = 25,
       units = "cm",
    res = 400)
plot(heatmap_all)
dev.off()

png(filename = "plots/heatmap_significant_LFC.png",
       width = 10,
       height = 25,
       units = "cm",
    res = 400)
plot(heatmap_sig)
dev.off()

png(filename = "plots/volcano_d0_d2.png",
       width = 16,
       height = 9,
       units = "cm",
    res = 400)
plot(heatmap_sig)
dev.off()
```
