---
title: "DESEQ_analysis"
author: "Sascha Maschmann"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

## Analysis of mito reads

After finding suitable scaling factors, we use these to normalize the mito mapped reads.  

```{r load data}
countFiles <- list.files("../out/counts",
                        pattern = "*mito_featureCounts.txt$",
                        full.names = TRUE)

mitoCounts <- read.delim(countFiles[1], header = T, skip = 1) %>% 
  select(Geneid, last_col())
for (file in countFiles[2:length(countFiles)]){
  counts <- read.delim(file, header = T, skip = 1) %>% 
    select(Geneid, last_col())
  mitoCounts <- left_join(mitoCounts, counts, by = "Geneid")
}


# Filter out all genes with less than 10 reads in all samples
cleanCounts <- mitoCounts %>% 
  filter(if_all(-Geneid, ~ .x >= 25))

colnames(cleanCounts) <- base::gsub("out.filtered.",
                                         "",
                                         colnames(cleanCounts))
colnames(cleanCounts) <- base::gsub("_mito_filtered.bam",
                                         "",
                                         colnames(cleanCounts))
colnames(cleanCounts) <- base::gsub("\\.",
                                         "_",
                                         colnames(cleanCounts))
```

```{r}
testplot <- cleanCounts %>% 
  filter(Geneid %in% c("SSUB", "RNA10", "RNA33")) %>% 
  pivot_longer(cols = 2:16,names_to = "con") %>% 
  mutate(con = substr(con, 1,3)) %>% 
  group_by(Geneid, con)
ggplot(testplot,
       aes(x = con,
           y = value,
           color = Geneid)) +
  geom_point() +
  theme_bw() +
  scale_y_log10() +
  facet_wrap(~Geneid)

```
```{r}
absoluteNormalizators <- base::readRDS("absolute_norm_values.RDS") %>% select(-con_b)

relativeCounts <- cleanCounts %>% 
  pivot_longer(cols = 2:16,
               names_to = 'con',
               values_to = 'reads') %>% 
  left_join(absoluteNormalizators,
            by="con") %>% 
  mutate(relative_counts = reads / value,
         con_plot = substr(con, 1,3)) %>% 
  group_by(Geneid) %>% 
  mutate(min = min(relative_counts),
         max = max(relative_counts)) %>% 
  ungroup() %>% 
  mutate(norm_counts = (relative_counts - min) / (max - min))

GOIs = c("SSUB", "RNA10", "RNA32", "LSUF")

ggplot(filter(relativeCounts, Geneid %in% GOIs),
       aes(x = con_plot,
           y = norm_counts,
           color = Geneid)) +
  geom_point() +
  theme_bw() +
  facet_wrap(~Geneid, scales ="free")
```
```{r calculate LFCs}
logfoldChanges <- relativeCounts %>% 
  select(Geneid, con, norm_counts) %>% 
  mutate(con = substr(con, 1,3)) %>% 
  group_by(Geneid, con) %>% 
  summarise(exp = mean(norm_counts)) %>% 
  pivot_wider(id_cols = Geneid,
              names_from = con,
              values_from = exp) %>% 
  ungroup() %>% 
  mutate(par_pr0 = log2(pr0 / par),
         pr0_pr1 = log2(pr1 / pr0),
         pr0_pr2 = log2(pr2 / pr0),
         pr0_pr3 = log2(pr3 / pr0),
         ) %>% 
  select(Geneid, par_pr0, pr0_pr1, pr0_pr2, pr0_pr3)
  
```

```{r calculate Significance}

genes <- unique(relativeCounts$Geneid)

pvals = data.frame()
for (gene in genes){
  for (xy in c("par_pr0", "pr0_pr1", "pr0_pr2", "pr0_pr3")){
    x = substr(xy, 1,3)
    y = substr(xy,5, 7)
    pvals[gene, xy] <- t.test(
      filter(relativeCounts,
             Geneid == gene,
             con_plot == x)$norm_counts,
      filter(relativeCounts,
             Geneid == gene,
             con_plot == y)$norm_counts,
      alternative = "two.sided"
    )[["p.value"]]
  }
}

fdrs = data.frame(row.names = row.names(pvals))
for (xy in c("par_pr0", "pr0_pr1", "pr0_pr2", "pr0_pr3")){
  fdrs[, xy] <- p.adjust(pvals[,xy], method = "BH")
}


```

```{r combine data}
expressionData = list()
for (gene in genes){
  expressionData[[gene]] <- data.frame(LFC = logfoldChanges %>% 
                                         filter(Geneid == gene) %>%
                                         select(-Geneid) %>% 
                                         t(),
                                       pval = pvals %>% 
                                         rownames_to_column("Geneid") %>% 
                                         filter(Geneid == gene) %>%
                                         select(-Geneid) %>% 
                                         t(),
                                       FDR = fdrs %>% 
                                         rownames_to_column("Geneid") %>% 
                                         filter(Geneid == gene) %>%
                                         select(-Geneid) %>% 
                                         t())
    
}

```

```{r HEatmap}
lfc_matrix <- logfoldChanges %>% 
  column_to_rownames("Geneid") %>% 
  as.matrix()
heatmap(lfc_matrix)
```

